Service/FilterService.php

```diff
@@ -97,6 +114,16 @@ public function getUrlOfFilteredImage($path, $filter, $resolver = null)
             $resolver
         );
 
+        if ($this->webpGenerate) {
```

Comment:
Hello! 

It assumes that the cache for the previous image has not been created, so it required to remove all previous cache in order to generate the webp format. It's not a big deal but i think it show an implementation issue.

<details>
<summary>We can build paths to resolve first then creating the cache for all the paths</summary>
<p>

```
public function getUrlOfFilteredImage($path, $filter, $resolver = null, $generateWebp = false)
{
    return $this->generateCacheFor($path, $filter, $generateWebp, $resolver);
}

public function getUrlOfFilteredImageWithRuntimeFilters(
    $path,
    $filter,
    array $runtimeFilters = [],
    $resolver = null,
    $generateWebp = false
) {
    return $this->generateCacheFor($path, $filter, $generateWebp, $runtimeFilters, $resolver);
}

public function generateCacheFor($path, $filter, $generateWebp, array $runtimeFilters = [], $resolver = null)
{
    $pathsToResolve = [
        new PathToResolveConfiguration(
            $path,
            [
                'filter' => $runtimeFilters,
            ]
        )
    ];

    // I also suggest to have only webp as file extension
    if ($generateWebp === true) {
        $webpPath = str_replace(
            \pathinfo($path, PATHINFO_EXTENSION),
            'webp',
            $path
        );

        $pathsToResolve []= new PathToResolveConfiguration(
            $webpPath,
            [
                'quality' => $this->webpQuality,
                'format' => 'webp',
                'filter' => $runtimeFilters,
            ]);
    }

    foreach ($pathsToResolve as $pathToResolve) {
        if ($this->cacheManager->isStored($pathToResolve->path(), $filter, $resolver)) {
            return $this->cacheManager->resolve($pathToResolve->path(), $filter, $resolver);
        }

        $filteredBinary = $this->createFilteredBinary(
            $pathToResolve->path(),
            $filter,
            $pathToResolve->options()
        );

        $this->cacheManager->store(
            $filteredBinary,
            $pathToResolve->path(),
            $filter,
            $resolver
        );

        return $this->cacheManager->resolve($pathToResolve->path(), $filter, $resolver);
    }
}

private function createFilteredBinary($path, $filter, array $options)
{
    $binary = $this->dataManager->find($filter, $path);

    try {
        return $this->filterManager->applyFilter($binary, $filter, $options);
    } catch (NonExistingFilterException $e) {
        $message = sprintf('Could not locate filter "%s" for path "%s". Message was "%s"', $filter, $path, $e->getMessage());

        $this->logger->debug($message);

        throw $e;
    }
}

class PathToResolveConfiguration
{
    private $path;
    private $options;

    public function __construct($path, array $options)
    {
        $this->path = $path;
        $this->options = $options;
    }

    public function path()
    {
        return $this->path;
    }

    public function options()
    {
        return $this->options;
    }
}
```
</p>
</details>

But i don't like either the condition `if ($generateWebp === true)` in this service which is more a functionnality in certain situation, a better way can be to use the Decorator pattern over `Service/FilterService` which will add the behavior of build also the webp image.

This decorator will be used depending of the context, if parameter given in cli to generate also webp or in the http controller if the browser accept webp.

<details>
<summary>Can be this implementation</summary>
<p>
```php
class WebpFilterServiceDecorator
{
	/**
     * @var FilterService
     */
	private $filterService;

	.. delegate other public method of FilterService (Decorator pattern)

	public function getUrlOfFilteredImage($path, $filter, $resolver = null)
	{
		$cachedPath = $this->filterService->getUrlOfFilteredImage($path, $filter, $resolver);

		$this->generateCacheForWebP($path, $filter, [], $resolver);

		return $cachedPath;
	}

	public function getUrlOfFilteredImageWithRuntimeFilters(
	        $path,
	        $filter,
	        array $runtimeFilters = [],
	        $resolver = null,
	        $webp = false
	) {
		$cachedPath = $this->filterService->getUrlOfFilteredImage($path, $filter, $resolver);

		$this->generateCacheForWebP($path, $filter, $runtimeFilters, $resolver);

		return $cachedPath;
	}

	private function generateCacheForWebP($path, $filter, $runtimeFilters, $resolver)
	{

		$webpPath = str_replace(
            \pathinfo($path, PATHINFO_EXTENSION),
            'webp',
            $path
        );

	    $options = [
            'quality' => $this->webpQuality,
            'format' => 'webp',
            'filter' => $runtimeFilters,
        ];

		$this->filterService->generateCacheFor($path, $filter, $options, $resolver);
	}

}

class FilterService
{

	...


	public function getUrlOfFilteredImage($path, $filter, $resolver = null)
	{
		$options = [
			'filters' =>  [],
		];

	    return $this->generateCacheFor($path, $filter, $options, $resolver);
	}

	public function getUrlOfFilteredImageWithRuntimeFilters(
	    $path,
	    $filter,
	    array $runtimeFilters = [],
	    $resolver = null
	) {
		$options = [
			'filters' => $runtimeFilters,
		];

	    return $this->generateCacheFor($path, $filter, $options, $resolver);
	}

	public function generateCacheFor($path, $filter, $generateWebp, array $options, $resolver = null)
	{
	  
        if ($this->cacheManager->isStored($path, $filter, $resolver)) {
            return $this->cacheManager->resolve($path, $filter, $resolver);
        }

        $filteredBinary = $this->createFilteredBinary(
            $path,
            $filter,
            $options
        );

        $this->cacheManager->store(
            $filteredBinary,
            $path,
            $filter,
            $resolver
        );

        return $this->cacheManager->resolve($pathToResolve->path(), $filter, $resolver);
	}

	private function createFilteredBinary($path, $filter, array $options = [])
    {
        $binary = $this->dataManager->find($filter, $path);

        try {
            return $this->filterManager->applyFilter($binary, $filter, $options);
        } catch (NonExistingFilterException $e) {
            $message = sprintf('Could not locate filter "%s" for path "%s". Message was "%s"', $filter, $path, $e->getMessage());

            $this->logger->debug($message);

            throw $e;
        }
    }
}

```
</p>
</details>

Reply:
A decorator imposes complexity in configuration and can increase the number of points of BC breaks.
And yes, i agree that this example is not very optimal.
Our project uses a different approach for generating WebP.

Reply:
You forgot one more condition.  WebP version of the image will be returned not only if WebP generation is enabled, but the browser also supports WebP. We get this value through the method arguments, and if we make a decorator for WebP, then this argument will be useless in `FilterService`.  That is, we cannot use the decorator pattern here.

> I also suggest to have only webp as file extension

Your code may overwrite existing files.
In my example, the original file extension becomes part of the file name. I don't see any problems in adding a suffix, but there may be problems with changing the file extension.

> It assumes that the cache for the previous image has not been created, so it required to remove all previous cache in order to generate the webp format.

Thanks for this comment.  I will fix this problem.